core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -383,6 +385,12 @@ private[log] class Cleaner(val id: Int,
       // trim excess index
       index.trimToValidSize()
 
+      // Append the last index entry
+      cleaned.timeIndex.maybeAppend(cleaned.largestTimestamp, segments.last.timeIndex.lastEntry.offset)
```

Comment:
Hmm, is this right? The offset associated with the largest timestamp may not be the last offset in the time index.